tests/common/dualtor/dual_tor_utils.py

```diff
@@ -667,54 +671,112 @@ def check_tunnel_balance(ptfhost, standby_tor_mac, vlan_mac, active_tor_ip, stan
+
+        inner_packet = send_pkt[IPv6].copy()
+        inner_packet[IPv6].hlim -= 1
+        exp_tunnel_pkt = testutils.simple_ipv4ip_packet(
+            eth_dst=dst_mac,
+            eth_src=src_mac,
+            ip_src="10.1.0.32",
+            ip_dst="10.1.0.33",
+            inner_frame=inner_packet
+        )
+        exp_tunnel_pkt = mask.Mask(exp_tunnel_pkt)
+        exp_tunnel_pkt.set_do_not_care_scapy(scapyall.Ether, "dst")
+        exp_tunnel_pkt.set_do_not_care_scapy(scapyall.Ether, "src")
+        exp_tunnel_pkt.set_do_not_care_scapy(scapyall.IP, "id")
+        exp_tunnel_pkt.set_do_not_care_scapy(scapyall.IP, "ttl")
+        exp_tunnel_pkt.set_do_not_care_scapy(scapyall.IP, "chksum")
+
+        return send_pkt, exp_pkt, exp_tunnel_pkt
+
     src_mac = ptfadapter.dataplane.get_mac(0, 0)
-    ip_dst = target_server_ip
-    SRC_IP_RANGE = [unicode('1.0.0.0'), unicode('200.255.255.255')]
-    ip_src = random_ip(SRC_IP_RANGE[0], SRC_IP_RANGE[1]) if 'src-ip' in hash_key else SRC_IP_RANGE[0]
-    sport = random.randint(1, 65535) if 'src-port' in hash_key else 1234
-    dport = random.randint(1, 65535) if 'dst-port' in hash_key else 80
     dst_mac = duthost.facts["router_mac"]
-    send_pkt = testutils.simple_tcp_packet(pktlen=128,
-                        eth_dst=dst_mac,
-                        eth_src=src_mac,
-                        dl_vlan_enable=False,
-                        vlan_vid=0,
-                        vlan_pcp=0,
-                        ip_src=ip_src,
-                        ip_dst=ip_dst,
-                        tcp_sport=sport,
-                        tcp_dport=dport,
-                        ip_ttl=64)
-    exp_pkt = mask.Mask(send_pkt)
-    exp_pkt.set_do_not_care_scapy(scapyall.Ether, 'dst')
-    exp_pkt.set_do_not_care_scapy(scapyall.Ether, "src")
-    exp_pkt.set_do_not_care_scapy(scapyall.IP, "ttl")
-    exp_pkt.set_do_not_care_scapy(scapyall.IP, "chksum")
 
-    inner_packet = send_pkt[IP]
-    inner_packet.ttl = inner_packet.ttl - 1
-    exp_tunnel_pkt = testutils.simple_ipv4ip_packet(
-        eth_dst=dst_mac,
-        eth_src=src_mac,
-        ip_src="10.1.0.32",
-        ip_dst="10.1.0.33",
-        inner_frame=inner_packet
-    )
-    send_pkt.ttl = 64
-    exp_tunnel_pkt[TCP] = inner_packet[TCP]
-    exp_tunnel_pkt = mask.Mask(exp_tunnel_pkt)
-    exp_tunnel_pkt.set_do_not_care_scapy(scapyall.Ether, "dst")
-    exp_tunnel_pkt.set_do_not_care_scapy(scapyall.Ether, "src")
-    exp_tunnel_pkt.set_do_not_care_scapy(scapyall.IP, "id") # since src and dst changed, ID would change too
-    exp_tunnel_pkt.set_do_not_care_scapy(scapyall.IP, "ttl") # ttl in outer packet is set to 255
-    exp_tunnel_pkt.set_do_not_care_scapy(scapyall.IP, "chksum") # checksum would differ as the IP header is not the same
+    # initialize the packets cache
+    if not hasattr(generate_hashed_packet_to_server, "packets_cache"):
+        generate_hashed_packet_to_server.packets_cache = defaultdict(list)
```

Comment:
Does this require a corresponding `from collections import defaultdict`?

Reply:
Yes, check out LINE#17

Reply:
My bad, didn't realize it was already imported!